bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java

```diff
@@ -1067,7 +1067,6 @@ public void asyncAddEntry(final byte[] data, final int offset, final int length,
     }
 
     public void asyncAddEntry(ByteBuf data, final AddCallback cb, final Object ctx) {
-        data.retain();
```

Comment:
one thing I am not very clear at this change is : 

if the change is move `data.retain` to PendingAddOp.create, why do we need to changes below?

- why do we need to do `ReferenceCountUtil.release(payload);` in PendingAddOp?
- why do we need to release in DigestManager?

I am a bit confused now. My biggest worry is if we need to change dlog code, that might mean it might potentially break other bookkeeper client.

Can you clarify the change here? I feel I am missing something here.

Reply:
@sijie  @dlg99 
I think it is safer to not change PendingAddOp and DigestManager.
We only need to add a data.retain() in the asyncAddEntry private method that Andrey is changing to public, nothing more.
The change that I am proposing will not touch any other code path.

Current code is running very well with v2 protocol on the LedgerHandle (non-Adv I mean) and so we are sure that reference counting is done very well.
in the case of LedgerHandleAdv we never exposed the ByteBuf API and we always used ByteBuf which were wrappers for byte[] so they are not reference counted.

Reply:
This thing grew a bit out of "let's make method public" change ;)

I'll start with DigestManager:
it handles V2 and V3 protocol.
For V2 protocol it returns ByteBufList which wraps received ByteBuf data and needs it until ByteBufList is released.
ByteBufList (aka toSend object in PendingAddOp) is released when callback is triggered. On ByteBufList release it callse .release() on all its parts, including ByteBuf data.

For V3 protocol currently it copies content into new unpooled ByteBuf, original data is no longer needed and no longer used so it should be released. This keeps behavior consistent for v2 and v3 in regards of ownership expectations of data (as in: computeDigestAndPackageForSending gets ownership of 1 refCnt of the passed ByteBuf and releases it when data is no longer needed or passes to a component that will do it).

This matches netty's "general rule of thumb is that the party who accesses a reference-counted object lastly is responsible for the destruction of the reference-counted object."

PendingAddOp keeps reference to payload.
Looking closer at it, there is no need to retain data unless we decide to i.e. log it after callback triggered and toSend was released. I'll actually remove it and re-run bookie tests locally, probably will push the change.
in this case payload's release in PendingAddOp should happen in recycle() only if payload was never used (hasRun == false). 

If I understood correctly, retain() call was added without accompanying release to fix failing test(s), i.e. DeferredSyncTest. This works with i.e. buffers produced by Unpooled.wrappedBuffer(..) but not when one tries to use Pooled allocator where unreleased reference will lead to a leak. I approached it by changing the test.

I understand your concerns about dlog. I propose the following:

1. Let's agree on the ownership rules for the ByteBufs and fix the client to use these rules consistently.

2. Let's review dlog's code. Any API calls in write path that pass ByteBuf directly are suspects, as well as parts where .slice() or .duplicate() is called. Tests caught one place where ref count had to be incremented. The rule of thumb is if ByteBuf is used after being passed somewhere, it has to be retained.
I am not familiar with dlog's codebase so I'll need help. We can add asserts on recCnt to existing tests, similar to ones in the new tests I added.
Existing tests caught few cases of reusing released ByteBuf (EnvelopedEntryWriter.getBuffer() ) and I did not find any other suspects after brief look at the dlog's code.
We can try enabling paranoid leak detection in unit tests, but at least initially it will catch false positives related to i.e. keeping references like
static final ByteBuf DATA = Unpooled.wrappedBuffer("foobar".getBytes()); 
and never releasing them at the end of the test. These won't be hard to fix but tedious.

3. We can add a warning to release notes.

Reply:
@dlg99 thank you for taking care of this.

I took a deeper look ad PendingAddOp#payload.
It should release *payload* after computeDigestAndPackageForSending and as you are saying it should release *payload* in case of hasRun == false.

So in order to make it clearer I suggest adding release in PendingAddOp and not in DigestManager. This way it is clear how  PendingAddOp is changing reference counts.

Reply:
@dlg99 

I haven't fully read your reply yet. but just to clarify what I was saying.

I uses dlog as an example. I am kind of less worried about dlog, since it is in the same repo and have all the tests running to make sure the change is correct at some extend. 

I am more worried about the applications that use bk that isn't in this repo, for example pulsar. it is a bit hard to tell what would the impact to those applications. it sounds like those applications have to change their behavior around reference counting. That's why I was worried this change become a behavior breaking change.

Reply:
@sijie one way or another we break something. 
We had inconsistency between LH and LHAdv.
This is new API overall and maintaining bug-compatibility feels like an overhead. Fixing it later won't be easier.
I'd love to make it consistent between LH and LHAdv, and consistent with netty's recommendations.
as alternative, I can slap .retain() call back into LH.asyncAddEntry to maintain that compatibility.
I'd very much like to not have to use it with LHAdv's variant because it would mean that I need to remember to release buffer (as in: try/finally around everything) after each call unlike the rest of the netty calls. But that means keeping inconsistency which will bite us eventually.

I suggest we sleep on it and decide tomorrow.
I'll look at Enrico's and your comments tomorrow and will address everything that does not touch subject of ref-cnt ownership.
Maybe we need to chat, like in slack tomorrow or I'll do my best to call in to community meeting this Thursday.

Reply:
@dlg99 The main reason for doing `retain()` from within `asyncAddEntry()` was to have a "friendly API". 

In general, the contract with Netty ByteBufs is that if some API is going to take temporarely ownership of a buffer, it should retain (and later release() accordingly). 

In this case, the retain is to make sure the buffer is valid until the end of the write operation, irrespective of what the BK app is doing with the buffer. 

For example, without this retain, an application should make sure the buffer is valid till the end, thus doing something like: 

```java
data.retain();
ledger.asyncAddEntry(data, (rc, entryId) -> {
    // do something 
    data.release();
}, ctx);
```

This would just force the application to do the same. 

The problem with the ByteBufList that is currently computed is that this happens from the ordered exectutor and not from the same thread that is calling the `asyncAddEntry()`, thus the buffer might be already invalid if application wouldn't call retain() itself.

Reply:
@merlimat there are a few discussions around this in the CR.

my counterarguments:

1. netty's http://netty.io/wiki/reference-counted-objects.html#who-destroys-it 
"The general rule of thumb is that the party who accesses a reference-counted object lastly is responsible for the destruction of the reference-counted object."

2. add's callback does not pass data as a parameter, does not need to access it, so no need to release within it as long as there is no retain().

This would do:
```java
ledger.asyncAddEntry(data, (rc, entryId) -> {
    // do something 
}, ctx);
```

now, if app needs data after addition:
```java
data.retain();
try {
  ledger.asyncAddEntry(data, (rc, entryId) -> {
      // do something 
  }, ctx);
  ...
  log.info("data was {}", asHex(data));
} finally {
   data.release()
}
```

alternative, with data.retain happening within asyncAddEntry one would have to do something like:

```java
try {
  ledger.asyncAddEntry(data, (rc, entryId) -> {
      // do something 
  }, ctx);
} finally {
  data.release();
}
```
or do that in the callback, as in your example.

Optionally we can follow what netty is doing with some of ByteBuf's methods:
have asyncAddEntry(ByteBuf...)
and asyncAddEntryAndRetain(ByteBuf...)
with well documented differences.

Note that all these data retention things only apply to methods that use netty's ByteBuf directly, nothing changes when byte[] on java's ByteBuffer variant used.

Reply:
@merlimat please look at the unit tests that I added
specifically assert for refcount https://github.com/apache/bookkeeper/pull/1585/files#diff-d168699d534345c67678296447f8375cR1344
has to pass (refcount after add == 0).
Otherwise it means that:
- whoever created ByteBuf has to deal with its release, despite netty's guidelines.
- if not released we risk memleak.

Reply:
Hmm, netty does have weird recommendation practice. :)